### PR TITLE
[FLINK-36249][streaming-java] Remove RestartStrategy-related configuration getters/setters that return/set complex Java objects.

### PR DESCRIPTION
**Pull request by JunRuiLee**
## What is the purpose of the change

[FLINK-33581](https://issues.apache.org/jira/browse/FLINK-33581)/FLIP-381 Deprecate configuration getters/setters that return/set complex Java objects. 
In Flink 2.0 we should remove these deprecated method and fields. This change will prevent users from configuring their jobs by passing complex Java objects, encouraging them to use ConfigOption instead.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by JunRuiLee** [comment]
@flinkbot run azure

**Comment by JunRuiLee** [comment]
@flinkbot run azure

**Comment by JunRuiLee** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java [hunk 1]
```diff
@@ -74,7 +76,14 @@ public void testEnumeratorCreationFails() throws Exception {
...
 
         final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
-        env.setRestartStrategy(RestartStrategies.fixedDelayRestart(Integer.MAX_VALUE, 0L));
+        Configuration configuration = new Configuration();
+        configuration.set(RestartStrategyOptions.RESTART_STRATEGY, "fixeddelay");
+        configuration.set(
+                RestartStrategyOptions.RESTART_STRATEGY_FIXED_DELAY_ATTEMPTS, Integer.MAX_VALUE);
+        configuration.set(
+                RestartStrategyOptions.RESTART_STRATEGY_FIXED_DELAY_DELAY, Duration.ofMillis(0));
+        env.configure(configuration, Thread.currentThread().getContextClassLoader());
```
Comment (zhuzhurk) [inline flink-connectors/flink-connector-base/src/test/java/org/apache/flink/connector/base/source/reader/CoordinatedSourceITCase.java:85]: Is it possible introduce some util methods to make it simpler? e.g. setFixedDelayRestartStrategy(env, attempt, delay).

File: flink-core/src/main/java/org/apache/flink/api/common/ArchivedExecutionConfig.java [hunk 2]
```diff
@@ -44,11 +44,10 @@ public class ArchivedExecutionConfig implements Serializable {
...
     public ArchivedExecutionConfig(ExecutionConfig ec) {
         executionMode = ec.getExecutionMode().name();
-        if (ec.getRestartStrategy() != null) {
-            restartStrategyDescription = ec.getRestartStrategy().getDescription();
-        } else {
-            restartStrategyDescription = "default";
-        }
+        // After remove restart strategy public API, we should set the restart strategy description
+        // is always "Cluster level default restart strategy" for backward compatibility
+        restartStrategyDescription = "Cluster level default restart strategy";
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/api/common/ArchivedExecutionConfig.java:49]: Looks to me "Cluster level default restart strategy" is not correct if users set restart strategy in Flink application through configuration.

File: flink-core/src/main/java/org/apache/flink/api/common/ArchivedExecutionConfig.java [hunk 3]
```diff
@@ -104,4 +105,80 @@ public long getPeriodicMaterializeIntervalMillis() {
     public Map<String, String> getGlobalJobParameters() {
         return globalJobParameters;
     }
+
+    private String getRestartStrategyDescription(Configuration configuration) {
```
Comment (zhuzhurk) [inline flink-core/src/main/java/org/apache/flink/api/common/ArchivedExecutionConfig.java:109]: Can we introduce an internal util class to host these new codes?
And add some tests for it.

File: flink-runtime/src/main/java/org/apache/flink/streaming/util/RestartStrategyUtils.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.RestartStrategyOptions;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+
+import java.time.Duration;
+
+/** Utility class for configuring restart strategies. */
+public class RestartStrategyUtils {
```
[the remark this answers is not in the document]
Comment (JunRuiLee) [reply flink-runtime/src/main/java/org/apache/flink/streaming/util/RestartStrategyUtils.java:29]: It's not easy to make this a test utility, as several components depend on it, including DataStreamAllroundTestJobFactory, and it isn't located in the test package.

[the remark this answers is not in the document]
Comment (JunRuiLee) [reply flink-runtime/src/main/java/org/apache/flink/streaming/util/RestartStrategyUtils.java:29]: I'll remove the usage of RestartStrategyUtils in the end-to-end cases and use RestartStrategyOptions instead.

**Comment by JunRuiLee** [comment]
@flinkbot run azure

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request closed without merging**